### PR TITLE
[Merged by Bors] - feat(frontends/lean/pp): add option pp.parens to force parentheses for notation

### PR DESCRIPTION
**Pull request by kmill**
With `set_option pp.parens true`, the pretty printer wraps all notation in parentheses regardless of precedence. For clarity, it wraps negative numerals in parentheses as well. However, it does not affect parentheses in universe level expressions.

For example,
```
#check @nat.add_assoc
/- nat.add_assoc : ∀ (n m k : ℕ), (((n + m) + k) = (n + (m + k))) -/
```

### CODE REVIEW

**Comment by fpvandoorn** [comment]
Does it put parentheses around notation with binders? Around the `foldr`-notation combinator (like for `[1,2,3]`)?
Do we want to add a test?

**Comment by kmill** [comment]
@fpvandoorn I added some tests to illustrate the behavior. It puts parentheses around every instance of non-special-cased notation (and also around negative numerals, just for clarity).

The only notation it does not parenthesize right now are ones built into the pretty printer, like subtypes, set separation, set of, explicit collections, tuples, and special constants like unit. It also does not parenthesize other constructions, like lambdas, function applications, pi expressions, let expressions, etc.

I think the main intent of this feature is to help understand precedence of user-defined notation.

**Comment by fpvandoorn** [comment]
LGTM

File: tests/lean/pp_parens.lean.expected.out [hunk 1]
```diff
@@ -0,0 +1,8 @@
+((1 + (2 * 3)) + 4) : ℕ
+nat.add_assoc : ∀ (n m k : ℕ), (((n + m) + k) = (n + (m + k)))
+(([1, 2, 3]) ++ ([3, 4, 5])) : list ℕ
+∀ (x : ℤ), ℤ → ℤ → ((x - (-3)) = ((-2) + x)) : Prop
+∀ (p : ℕ → ℕ → Prop), p (1 + 2) 3 : Prop
+∀ (f : ℕ → ℕ → ℕ), ((f 2 (f 3 4 + 1) + 2) = f 1 2) : Prop
```
Comment (digama0) [inline tests/lean/pp_parens.lean.expected.out:6]: Shouldn't there be parentheses around `f 3 4` and `f 1 2` in this example? It's not always obvious whether a notation has precedence higher than application or not.

Comment (kmill) [reply tests/lean/pp_parens.lean.expected.out:6]: @digama0 Here's an example for how high-precedence notation interacts with the feature:
```lean
local infix ` ++ `:5000 := list.append

def tail {α : Type*} : list α → list α := list.tail

#check (tail [1,2,3])++[3,4,5]
/- (tail [1, 2, 3]) ++ [3, 4, 5] : list ℕ -/
#check tail [1,2,3]++[3,4,5]
/- tail [1, 2, 3] ++ [3, 4, 5] : list ℕ -/

set_option pp.parens true

#check (tail [1,2,3])++[3,4,5]
/- ((tail ([1, 2, 3])) ++ ([3, 4, 5])) : list ℕ -/
#check tail [1,2,3]++[3,4,5]
/- tail (([1, 2, 3]) ++ ([3, 4, 5])) : list ℕ -/
```
It seems like everything's unambiguous, even if you don't know any of the precedences. (Or, at least, if you assume function application has highest precedence.)

I had looked a little into having function applications *always* be wrapped in an additional layer of parentheses, but at least in the few easy-ish ways I'd thought of implementing it, you'd get things like `(((f) 3) 4)` or `((f x y) + (x))`.

**Pull request closed without merging**